indico/MaKaC/common/output.py

```diff
@@ -957,234 +940,228 @@ def _confToXMLMarc21(self,conf,includeSession=1,includeContribution=1,includeMat
+        out.writeTag("subfield", 'INDICO.{}'.format(uniqueId(contrib.event_new)), [["code", "b"]])
         out.closeTag("datafield")
 
         out.openTag("datafield",[["tag","970"],["ind1"," "],["ind2"," "]])
-        confcont = "INDICO." + uniqueId(cont)
-        out.writeTag("subfield",confcont,[["code","a"]])
+        out.writeTag("subfield", 'INDICO.{}'.format(uniqueId(contrib)), [["code", "a"]])
         out.closeTag("datafield")
 
         out.openTag("datafield",[["tag","980"],["ind1"," "],["ind2"," "]])
-        out.writeTag("subfield", self._getRecordCollection(cont), [["code","a"]])
+        out.writeTag("subfield", self._getRecordCollection(contrib), [["code","a"]])
         out.closeTag("datafield")
 
-        self._generateLinkField(urlHandlers.UHContributionDisplay,
-                                cont, "Contribution details", out)
-        self._generateLinkField(urlHandlers.UHConferenceDisplay,
-                                cont.getConference(), "Event details", out)
+        # TODO: Add proper link to contribution display page
+        self._generate_link_field('#', 'Contribution details', out)
 
-        self._generateAccessList(cont, out, objId=uniqueId(cont))
+        self._generate_link_field(url_for('event.conferenceDisplay', confId=contrib.event_new.id, _external=True),
+                                  'Event details', out)
+
+        self._generateAccessList(contrib, out, objId=uniqueId(contrib))
     ####
     #fb
 
+    def _generate_link_field(self, url, caption, out):
+        out.openTag('datafield', [['tag', '856'], ['ind1', '4'], ['ind2', ' ']])
+        out.writeTag('subfield', url, [['code', 'u']])
+        out.writeTag('subfield', caption, [['code', 'y']])
+        out.closeTag('datafield')
+
+    def _generate_references(self, obj, out):
+        if obj.references:
+            out.openTag('datafield', [['tag', '088'], ['ind1', ' '], ['ind2', ' ']])
+            for reference in obj.references:
+                out.writeTag('subfield', reference.value, [['code', 'a']])
+            out.closeTag('datafield')
+
+    def _generate_contrib_location_and_time(self, contrib, out):
+        if contrib.venue_name or contrib.start_dt:
+            out.openTag('datafield', [['tag', '518'], ['ind1', ' '], ['ind2', ' ']])
+            if contrib.venue_name:
+                out.writeTag('subfield', contrib.venue_name, [['code', 'r']])
+            if contrib.start_dt is not None:
+                out.writeTag('subfield', contrib.start_dt.strftime('%Y-%m-%dT%H:%M:00Z'), [['code', 'd']])
+                out.writeTag('subfield', contrib.end_dt.strftime('%Y-%m-%dT%H:%M:00Z'), [['code', 'h']])
+            out.closeTag('datafield')
+
+    def _generate_contrib_people(self, contrib, out, subcontrib=None):
+        aList = list(contrib.primary_authors | contrib.secondary_authors)
+        sList = list(subcontrib.person_links) if subcontrib else list(contrib.speakers)
+        list_ = {}
```

Comment:
that's a pretty awful name, especially since it's not even a list but a dict ;)
also, good candidate for a defaultdict. the code below can probably be cleaned up a lot too